### PR TITLE
BSplineInterpolation constructors that take the coefficient matrix

### PR DESCRIPTION
**Pull request by nlw0**
Patch for my own feature request #284

### CODE REVIEW

**Comment by timholy** [comment]
Looks good. Can you add a test? As future developers update the package, having a test will ensure this won't silently break---if it causes your test(s) to fail, it will force him/her to figure out how to fix the problem before continuing.

**Comment by nlw0** [comment]
I can sure try, but I'm a total Julia newbie and appreciate any guidance.

Does the code look good? I'm not even familiar with using OffsetArrays, for instance. Could it be better if I find a way to define another method instead of having these multiple class constructors?

File: test/b-splines/cubic.jl [hunk 1]
```diff
@@ -68,4 +68,8 @@
     itp_flat_c = interpolate(A, BSpline(Cubic(Flat(OnCell()))))
     @test Interpolations.gradient(itp_flat_c,0.5)[1] ≈ 0 atol=eps()
     @test Interpolations.gradient(itp_flat_c,ix[end] + 0.5)[1] ≈ 0 atol=eps()
+
+    # Can construct from coefficients
+    itp = interpolate(rand(5, 9), BSpline(Cubic(Flat(OnGrid()))))
+    @test itp == Interpolations.BSplineInterpolation(itp.coefs, itp.it, itp.parentaxes)
```
Comment (timholy) [inline test/b-splines/cubic.jl]: Need to test the non-OffsetArrays constructor too. (If it doesn't work as intended it's better not to have it at all.)

**Comment by timholy** [comment]
> Does the code look really good? I'm not even familiar with using OffsetArrays, for instance. Could it be better if I find a way to define another method instead of having these multiple class constructors?

For an intro to OffsetArrays, see [this blog post](https://julialang.org/blog/2017/04/offset-arrays). It was written for an older version of Julia, so some info might be a bit outdated but it should give you the main points.

**Comment by nlw0** [comment]
I think the second constructor might be handy... I included a test for it, and also refactored the code a little bit

**Comment by nlw0** [comment]
I take that back, let's keep it simple

**Comment by nlw0** [comment]
@timholy Maybe the function could actually even take an AbstractArray instead? It's really just about avoiding the type argument, and any errors related to non-matching array ranges are just the user's fault here...

File: src/b-splines/b-splines.jl [hunk 2]
```diff
@@ -62,6 +62,10 @@ function BSplineInterpolation(::Type{TWeights}, A::AbstractArray{Tel,N}, it::IT,
     BSplineInterpolation{T,N,typeof(A),IT,typeof(axs)}(A, fix_axis.(axs), it)
 end
 
+function BSplineInterpolation(A::AbstractArray{Tel,N}, it::IT, axs) where {N,Tel,IT<:DimSpec{BSpline}}
```
Comment (timholy) [inline src/b-splines/b-splines.jl]: We need to do some sanity-checking:
```suggestion
function BSplineInterpolation(A::AbstractArray{Tel,N}, it::IT, axs) where {N,Tel,IT<:DimSpec{BSpline}}
    @noinline err_axes(A, it, axs) = throw(ArgumentError("parentaxes $axs are inconsistent with coefficient array axes $(axes(A)) and interpolation type $it"))

    paxs = padded_axes(axs, it)
    all(map(⊆, paxs, axes(A))) || err_axes(A, it, axs)
```

File: test/b-splines/cubic.jl (hunk 1 above)
Comment (timholy) [inline test/b-splines/cubic.jl]:
```suggestion
    @test itp == Interpolations.BSplineInterpolation(itp.coefs, itp.it, itp.parentaxes)
    @test_throws ArgumentError Interpolations.BSplineInterpolation(itp.coefs, itp.it, (2:3, 2:10))
```

**Comment by nlw0** [comment]
Great, thanks!

**Comment by timholy** [comment]
Thanks @nlw0!

**Pull request merged**